eth/chains/base.py

```diff
@@ -412,29 +411,27 @@ def from_genesis(cls,
         """
         genesis_vm_class = cls.get_vm_class_for_block_number(BlockNumber(0))
 
-        account_db = genesis_vm_class.get_state_class().get_account_db_class()(
-            base_db,
-            BLANK_ROOT_HASH,
-        )
+        pre_genesis_header = BlockHeader(difficulty=0, block_number=-1, gas_limit=0)
```

Comment:
This seems a little weird to me, rather than making a fake object it might better to change `bulid_state`'s signature to accept a `state_root`, then you can pass the state root in directly.

Reply:
Yeah, I do agree this is a little funky. But adding an explicit API where `ExecutionContext` is `None` is a serious footgun (any calls to `state.timestamp` etc will crash hard). Really the only scenario where it makes sense to generate state via a state root and no execution context is when you know there won't be any need to access state besides the database, which should only happen in the genesis block. I was a little hesitant to build a custom API method just for that.

I guess we could set `ExecutionContext` to `None` here in the genesis builder, and add a giant warning for other people not to copy that code elsewhere. I like that a tiny bit less than this, but a bit more than adding a footgun API.

Any other alternatives you see?

Reply:
Ahh, somehow I missed that `header` was being used for more than just its state root. Sure, so we could do the `genesis_vm_class.get_state_class()(db, None, BLANK_ROOT_HASH)` thing, but that's bad form since state holds more than the account data.

Maybe the problem is that this function isn't really using a state! It's using a state which happens to wrap an AccountDB and that's why more objects are getting pulled in. Maybe this function alone could continue using AccountDB directly?

Reply:
Except it also gets and sets storage, which won't be in `account_db` as of #1735 :confused:

Reply:
At that point we are saying we want something like a `StateDB` that's distinct from `State` which really becomes something like the old `AccountDB` was...

Reply:
Or maybe the opposite, and we extract `ExecutionContext` out as something that's adjacent to `State` instead of embedded in it. Then if you want to execute, you run it through the `VM` which combines state with the execution context generated with the header in order to actually build a computation.

I'm starting to think something like:

- move `BaseState.execution_context` -> `BaseVM.execution_context`
- move `BaseState.get_computation` -> `BaseVM.get_computation` (which builds with its local execution context)
- rename `ExecutionContext` -> `BlockContext`
- Computation learns about `BlockContext`, like it knows about `TransactionContext`
- `BaseState` now only needs a db and state root to be generated

I'll capture this into an issue for follow-up since I don't think it belongs here. I will also note that this hacky code should be cleaned up when `State` is easier to initialize.

Reply:
For future readers here's the issue: https://github.com/ethereum/py-evm/issues/1758